File: ClientRuntimes/NodeJS/ClientRuntime.Azure/lib/pollingState.js
﻿// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information. 
'use strict';

var util = require('util');
var msRest = require('ms-rest');
var LroStates = require('./constants').LongRunningOperationStates;

/**
 * @class
 * Initializes a new instance of the PollingState class.
 * @constructor
 * @param {object} [response] - The response object to extract long
 * running operation status.
 * 
 * @param {number} retryTimeout - The timeout in seconds to retry on
 * intermediate operation results.
 */
function PollingState(resultOfInitialRequest, retryTimeout) {
  this._retryTimeout = retryTimeout;
  this.response = this.updateResponse(resultOfInitialRequest.response);
  this.request = resultOfInitialRequest.request;
  this.resource = resultOfInitialRequest.body;
  
  if (this.resource && this.resource.properties && this.resource.properties.provisioningState) {
    this.status = this.resource.properties.provisioningState;
  } else {
    switch (this.response.statusCode) {
      case 202:
        this.status = LroStates.InProgress;
        break;

      case 204:
      case 201:
      case 200:
        this.status = LroStates.Succeeded;
        break;

      default:
        this.status = LroStates.Failed;
        break;
    }
  }
}

/**
 * Gets timeout in milliseconds. 
 * @returns {number} timeout
 */
PollingState.prototype.getTimeout = function () {
  if (this._retryTimeout || this._retryTimeout === 0) {
    return this._retryTimeout * 1000;
  }
  if (this.response && this.response.headers['retry-after']) {
    return parseInt(this.response.headers['retry-after']) * 1000;
  }
  return 30 * 1000;
};

PollingState.prototype.updateResponse = function (response) {
  this.response = response;
  if (response && response.headers) {
    if (response.headers['azure-asyncoperation']) {
      this.azureAsyncOperationHeaderLink = response.headers['azure-asyncoperation'];
    }
    
    if (response.headers['location']) {
      this.locationHeaderLink = response.headers['location'];
    }
  }
  return this.response;
};

/**
 * Returns long running operation result.
 * @returns {object} HttpOperationResponse
 */
PollingState.prototype.getOperationResponse = function () {
  var result = new msRest.HttpOperationResponse();
  result.request = this.request;
  result.response = this.response;
  result.body = this.resource;
  return result;
};

/**
 * Returns an Error on operation failure.
 * @returns {object} Error
 */
PollingState.prototype.getCloudError = function (err) {
  var errMsg;
  if (err && err.message) {
    errMsg = util.format('Long running operation failed with error: \'%s\'.', err.message);  
  } else {
    errMsg = util.format('Long running operation failed with status: \'%s\'.', this.status);
  }
  
  var error = new Error(errMsg);
  error.body = this.response.body;
  error.request = this.request;
  error.response = this.response;
  return error;
};

module.exports = PollingState;

File: ClientRuntimes/Java/azure-client-runtime/src/main/java/com/microsoft/rest/PollingState.java
/**
 *
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for license information.
 *
 */

package com.microsoft.rest;

import com.fasterxml.jackson.databind.JsonNode;
import com.microsoft.rest.serializer.JacksonHelper;
import retrofit.Response;

import java.io.IOException;

/**
 * An instance of this class defines the state of a long running operation.
 *
 * @param <T> the type of the resource the operation returns.
 */
public class PollingState<T> {
    private Response response;
    private String status;
    private String azureAsyncOperationHeaderLink;
    private String locationHeaderLink;
    private Integer retryTimeout;
    private T resource;
    private CloudError error;

    public PollingState(ServiceResponse<T> response, int retryTimeout) throws ServiceException {
        this.retryTimeout = retryTimeout;
        this.setResponse(response.getResponse());
        this.resource = response.getBody();

        try {
            JsonNode resource = JacksonHelper.getObjectMapper().readTree(this.response.raw().body().byteStream());
            if (resource != null && resource.get("properties") != null &&
                    resource.get("properties").get("provisioningState") != null) {
                setStatus(resource.get("properties").get("provisioningState").asText());
            } else {
                switch (this.response.code()) {
                    case 202:
                        setStatus(AzureAsyncOperation.inProgressStatus);
                        break;
                    case 204:
                    case 201:
                    case 200:
                        setStatus(AzureAsyncOperation.successStatus);
                        break;
                    default:
                        setStatus(AzureAsyncOperation.failedStatus);
                }
            }
        } catch (Exception ex) {
            throw new ServiceException(ex);
        }
    }

    public int getDelayInMilliseconds() {
        if (this.retryTimeout != null) {
            return this.retryTimeout * 1000;
        }
        if (this.response != null) {
            return Integer.parseInt(response.headers().get("Retry-After"));
        }
        return AzureAsyncOperation.defaultDelay * 1000;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) throws IllegalArgumentException {
        if (status == null) {
            throw new IllegalArgumentException("Status is null.");
        }
        this.status = status;
    }

    public Response getResponse() {
        return this.response;
    }

    public void setResponse(Response response) {
        this.response = response;
        if (response != null) {
            this.azureAsyncOperationHeaderLink = response.headers().get("Azure-AsyncOperation");
            this.locationHeaderLink = response.headers().get("Location");
        }
    }

    public String getAzureAsyncOperationHeaderLink() {
        return azureAsyncOperationHeaderLink;
    }

    public String getLocationHeaderLink() {
        return locationHeaderLink;
    }

    public T getResource() {
        return resource;
    }

    public void setResource(T resource) {
        this.resource = resource;
    }

    public CloudError getError() {
        return error;
    }

    public void setError(CloudError error) {
        this.error = error;
    }
}
